bob-backend/src/pearl/postprocessor.rs

```diff
@@ -0,0 +1,109 @@
+use std::{
+    collections::BTreeSet,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use crate::{
+    pearl::{holder::PearlSync, Holder},
+    prelude::*,
+};
+
+#[derive(Clone)]
+pub(crate) struct SimpleHolder {
+    storage: Arc<RwLock<PearlSync>>,
+    timestamp: u64,
+}
+
+impl From<&Holder> for SimpleHolder {
+    fn from(holder: &Holder) -> Self {
+        let storage = holder.cloned_storage();
+        let timestamp = holder.end_timestamp();
+        Self { storage, timestamp }
+    }
+}
+
+impl SimpleHolder {
+    pub(crate) async fn filter_memory_allocated(&self) -> usize {
+        self.storage.read().await.filter_memory_allocated().await
+    }
+
+    pub(crate) async fn offload_filter(&self) {
+        self.storage.read().await.offload_filters().await
+    }
+
+    pub(crate) fn timestamp(&self) -> u64 {
+        self.timestamp
+    }
+
+    pub(crate) async fn is_ready(&self) -> bool {
+        self.storage.read().await.is_ready()
+    }
+}
+
+impl PartialOrd for SimpleHolder {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.timestamp.partial_cmp(&other.timestamp)
+    }
+}
+
+impl PartialEq for SimpleHolder {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.storage, &other.storage) && self.timestamp == other.timestamp
+    }
+}
+
+impl Ord for SimpleHolder {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.timestamp.cmp(&other.timestamp)
+    }
+}
+
+impl Eq for SimpleHolder {}
+
+#[derive(Clone, Default)]
+pub struct PostProcessor {
+    holders: Arc<RwLock<BTreeSet<SimpleHolder>>>,
+    allocated_size: Arc<AtomicUsize>,
+    filter_memory_limit: Option<usize>,
+}
+
+impl PostProcessor {
+    pub(crate) fn new(filter_memory_limit: Option<usize>) -> Self {
+        Self {
+            filter_memory_limit,
+            ..Default::default()
+        }
+    }
+    pub(crate) async fn storage_prepared(&self, holder: impl Into<SimpleHolder>) {
+        let mut holders = self.holders.write().await;
+        let holder = holder.into();
+        let filter_memory = holder.filter_memory_allocated().await;
+        holders.insert(holder);
+        self.allocated_size
```

Comment:
As I see, only this function have an access to `allocated_size` and it does not work in parallel. If I am right, `Atomic` here is overkill and can be replaced with a simple integer value

Reply:
But this structure is cloned to different dcs, groups and holders, and they can call this method in parallel

Reply:
Ok